### PR TITLE
`Development`: Modernize chart visualization of active students and performance in exercises

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
### Motivation and Context

This PR changes the line chart implementation of the _Active Students_ chart in each course card on the course management page and the _Performance in Exercises_ chart in the student statistics tab of every course from `Chart.js` to `ngx-charts`.

If you take a detailed look at both representations of _Active Students_, you might notice that the old one provides data labels displaying the absolute value for every point, which the new one doesn't. I'm aware of this difference, but wasn't able to find a possibility to implement this with `ngx-charts` so far.
### Description

The html templates as well as the underlying data processing logics in the components have been simplified and adapted in order to fit `ngx-charts` requirements.

### Steps for Testing

Prerequisites:
**Important:** before deploying this PR to a test server, take a look at the current implementation. This includes hovering over the line, adjusting screen size, clicking on nodes in the chart (_Performance in Exercises_ only) and clicking on parts of the legend (_Performance in Exercises_ only).

- 1 Instructor
- 1 Course with students (at least one) and exercises
- 1 student participating on this course

1. Log in to Artemis as instructor
2. Navigate to Course Administration
3. Inspect the new implementation of _Active Students_. Again, hover over the line and adjust the screen size. The chart should now be responsive to the screen width.
4. Log out and log in as student again
5. Click on the course card for the corresponding course and enter the statistics tab
6. Scroll down to the _Performance in Exercises_ chart and again, interact with it. Check whether the interaction is similar to the old implementation. The chart should be responsive to the screen width as well now.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

course-management-overview-statistics.component.ts: 66.67% | 90.32%

### Screenshots

**Old design of _Performance in Exercises_:**
image
**Old design of _Active Students_:**
image

**New design of _Performance in Exercises_:**
image
**New design of _Active Students_:**
image

### CODE REVIEW

File: src/main/webapp/app/course/manage/overview/course-management-card.component.html [hunk 1]
```diff
@@ -174,7 +174,7 @@ <h4 class="text-center no-exercises mt-3 fw-medium">{{ 'artemisApp.course.noExer
                 </div>
             </div>
             <div class="col-4 section-card statistics-card align-center">
-                <div class="section-content align-center">
+                <div class="align-center" style="width: 100%; height: 300px">
```
Comment (Malyuk-A) [inline src/main/webapp/app/course/manage/overview/course-management-card.component.html:177]: Please extract inline style to an CSS file or use bootstrap

File: src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.html [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
-        baseChart
-        height="300"
-        width="750"
-        [datasets]="chartData"
-        [labels]="barChartLabels"
-        [options]="barChartOptions"
-        [legend]="barChartLegend"
-        [chartType]="barChartType"
-    ></canvas>
+<div #containerRef class="col-md-12 text-center" style="max-height: 300px; justify-content: center; align-items: center">
```
Comment (Malyuk-A) [inline src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.html:2]: here as well

File: src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts [hunk 3]
```diff
@@ -79,46 +64,73 @@ export class ExerciseScoresChartComponent implements AfterViewInit, OnDestroy {
...
-        this.chartInstance.resize();
         // we show all the exercises ordered by their release data
         const sortedExerciseScores = sortBy(this.exerciseScores, (exerciseScore) => exerciseScore.releaseDate);
-        this.addData(this.chartInstance, sortedExerciseScores);
+        this.addData(sortedExerciseScores);
     }
 
-    private addData(chart: Chart, exerciseScoresDTOs: ExerciseScoresDTO[]) {
-        for (const exerciseScoreDTO of exerciseScoresDTOs) {
+    private addData(exerciseScoresDTOs: ExerciseScoresDTO[]) {
```
Comment (lschlesinger) [inline src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts:72]: you could add a doc here

File: src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts [hunk 4]
```diff
@@ -79,46 +64,73 @@ export class ExerciseScoresChartComponent implements AfterViewInit, OnDestroy {
...
+            this.navigateToExercise(data.exerciseId);
+        } else {
+            // if a legend label is clicked, the corresponding line has to disappear or reappear
+            const name = JSON.parse(JSON.stringify(data)) as string;
+            // find the affected line in the dataset
+            const index = this.ngxData.findIndex((dataPack) => {
+                const dataName = dataPack.name as string;
+                return dataName === name;
+            });
+            console.log(index);
```
Comment (lschlesinger) [inline src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts:116]: this should be removed?

File: src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.html [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
+        [xAxis]="true"
+        [yAxis]="true"
+        [yScaleMin]="0"
+        [yScaleMax]="100"
+        [yAxisTickFormatting]="formatYAxis"
+        [rotateXAxisTicks]="false"
+        [animations]="true"
+    >
+        <ng-template #tooltipTemplate let-model="model">
+            <h3 *ngIf="findAbsoluteValue(model) !== 1" class="tooltip-header">{{ findAbsoluteValue(model) }} active students</h3>
```
Comment (Malyuk-A) [inline src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.html:16]: Would be great to add translations here

File: src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts [hunk 6]
```diff
@@ -79,46 +64,73 @@ export class ExerciseScoresChartComponent implements AfterViewInit, OnDestroy {
...
+    onSelect(data: any) {
+        // delegate to the corresponding exericse if chart node is clicked
+        if (data.exerciseId) {
+            console.log(JSON.parse(JSON.stringify(data)));
+            this.navigateToExercise(data.exerciseId);
+        } else {
+            // if a legend label is clicked, the corresponding line has to disappear or reappear
+            const name = JSON.parse(JSON.stringify(data)) as string;
+            // find the affected line in the dataset
+            const index = this.ngxData.findIndex((dataPack) => {
```
Comment (lschlesinger) [inline src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts:112]: could you add a type here? `((dataPack: Type) => ...`

File: src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts [hunk 7]
```diff
@@ -79,46 +64,73 @@ export class ExerciseScoresChartComponent implements AfterViewInit, OnDestroy {
...
+    /**
+     * Provides the functionality when the user interacts with the chart by clicking on it.
+     * If the users click on a node in the chart, they get delegated to the corresponding exercise detail page.
+     * If the users click on an entry in the legend, the corresponding line disappears or reappears depending on its previous state
+     * @param data the event sent by the framework
+     */
+    onSelect(data: any) {
+        // delegate to the corresponding exericse if chart node is clicked
+        if (data.exerciseId) {
+            console.log(JSON.parse(JSON.stringify(data)));
```
Comment (lschlesinger) [inline src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts:106]: this should be removed?

File: src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.ts [hunk 8]
```diff
@@ -21,90 +17,32 @@ export class CourseManagementOverviewStatisticsComponent implements OnInit, OnCh
...
-    dataForSpanType: number[] = [];
+    lineChartLabels: string[] = [];
+
+    // ngx-data
+    ngxData: any[] = [];
+    chartColor: Color = {
+        name: 'vivid',
+        selectable: true,
+        group: ScaleType.Ordinal,
+        domain: ['rgba(53,61,71,1)'],
```
Comment (Malyuk-A) [inline src/main/webapp/app/course/manage/overview/course-management-overview-statistics.component.ts:32]: nitpick: Would be great to have a comment here or a dedicated variable so you would know what color this is

**Review by lschlesinger — COMMENTED** [review]
I found some minor things that should be improved (see comments).
On top of the comments, you could do:
`private addData(exerciseScoresDTOs: ExerciseScoresDTO[]) : void` for example

**Review by Malyuk-A — COMMENTED** [review]
Found some possible improvements, the rest looks good. Tested functionality during testing session on TS1. Works well.

**Review by MDK5 — APPROVED** [review]
Tested in testing session, charts look good!

File: src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts (hunk 3 above)
Comment (willergeorg) [reply src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.ts]: All done in 4ed25bd

**Review by Malyuk-A — DISMISSED** [review]
Tested on TS1 and looked at the code changes. Works as described and code looks good. 👍

**Review by lschlesinger — APPROVED** [review]
feedback is addressed and the functionality was tested during testing session on TS1. Good job, the charts look waaaaay better :)

**Review by cremertim — APPROVED** [review]
Tested on ts3. The new charts work as expected. Code lgtm:

I tested according to the testing steps and did not encounter any issues

**Comment by krusche** [comment]
There is a merge conflict now, please resolve it

**Pull request merged**